File: 98-hexagonal/hexagonal_demo.py
# Ngôn ngữ: Python
# Công dụng: Bản Python của cùng bài học — miền định nghĩa CỔNG, hạ tầng cung cấp BỘ NỐI,
# mọi phụ thuộc chỉ đi vào TRONG. Ba con bug: miền gọi thẳng hạ tầng nên không test được;
# cổng nói tiếng bộ nối; và "cấu trúc thư mục lục giác" mà chiều phụ thuộc vẫn ngược.
# Tại sao cần học: Java và C++ đảo ngược phụ thuộc bằng KẾ THỪA — bộ nối phải `implements`
# / kế thừa cổng, tức là hạ tầng vẫn phải import miền. Python có một thứ mạnh hơn hẳn và
# ít người dùng đúng: `typing.Protocol`. Với nó, bộ nối KHÔNG cần biết cổng tồn tại — một
# lớp viết ra từ trước, chưa bao giờ nghe tên miền của bạn, vẫn thoả cổng nếu nó có đúng
# phương thức. Phụ thuộc bằng 0 ở CẢ HAI CHIỀU, và bài này chứng minh điều đó bằng một
# lớp thư viện bên thứ ba không hề liên quan.

from typing import Optional, Protocol, runtime_checkable


# =====================================================================
# MIỀN — không biết CSDL, mạng, đồng hồ hay khung phần mềm nào tồn tại
# =====================================================================
class DonHang:
    def __init__(self, ma, ma_khach, tong_tien, tao):
        if tong_tien <= 0:
            raise ValueError("tổng tiền phải dương")
        self.ma, self.ma_khach, self.tong_tien, self.tao = ma, ma_khach, tong_tien, tao


# ---- CỔNG BỊ ĐIỀU KHIỂN (miền GỌI RA ngoài) ----
# Chú ý từ ngữ: "tìm theo mã", "lưu", "báo cho khách". Không có `Cursor`, không có
# `Response`. Cổng nói tiếng NGHIỆP VỤ (bài 81).
@runtime_checkable
class KhoDonHang(Protocol):
    def tim_theo_ma(self, ma: str) -> Optional[DonHang]: ...
    def luu(self, d: DonHang) -> None: ...


@runtime_checkable
class BaoChoKhach(Protocol):
    def bao(self, ma_khach: str, noi_dung: str) -> None: ...


@runtime_checkable
class DongHo(Protocol):
    def bay_gio(self) -> int: ...            # bài 67


# ---- CỔNG ĐIỀU KHIỂN (thế giới GỌI VÀO miền) ----
@runtime_checkable
class DatHang(Protocol):
    def thuc_hien(self, ma_khach: str, tong_tien: int) -> DonHang: ...


class DichVuDatHang:
    """Lõi ứng dụng: cài cổng điều khiển, dùng cổng bị điều khiển. 0 phụ thuộc hạ tầng."""

    def __init__(self, kho: KhoDonHang, bao: BaoChoKhach, dong_ho: DongHo):
        self._kho, self._bao, self._dong_ho = kho, bao, dong_ho
        self._dem = 0

    def thuc_hien(self, ma_khach, tong_tien):
        self._dem += 1
        d = DonHang(f"DH-{self._dem}", ma_khach, tong_tien, self._dong_ho.bay_gio())
        self._kho.luu(d)
        self._bao.bao(ma_khach, f"đã tạo đơn {d.ma}")
        return d


# =====================================================================
# BỘ NỐI — hạ tầng. Chú ý: KHÔNG lớp nào kế thừa gì cả.
# =====================================================================
class KhoTrongBoNho:
    def __init__(self):
        self.bang = {}

    def tim_theo_ma(self, ma):
        return self.bang.get(ma)

    def luu(self, d):
        self.bang[d.ma] = d


class KhoSql:
    """Bộ nối thứ hai: giả lập SQL. Cùng cổng, cài đặt hoàn toàn khác."""

    def __init__(self):
        self.cau_lenh, self.bang = [], {}

    def tim_theo_ma(self, ma):
        self.cau_lenh.append(f"SELECT * FROM don_hang WHERE ma = '{ma}'")
        return self.bang.get(ma)

    def luu(self, d):
        self.cau_lenh.append(f"INSERT INTO don_hang VALUES ('{d.ma}', ...)")
        self.bang[d.ma] = d


class BaoGia:
    def __init__(self):
        self.da_bao = []

    def bao(self, ma_khach, noi_dung):
        self.da_bao.append(f"{ma_khach}:{noi_dung}")


class DongHoCoDinh:
    def __init__(self, luc):
        self.luc = luc

    def bay_gio(self):
        return self.luc


# =====================================================================
# BẢN SAI — miền gọi thẳng hạ tầng
# =====================================================================
class KetNoiCsdl:
    so_lan_mo_ket_noi = 0
    co_san = False

    def __init__(self):
        KetNoiCsdl.so_lan_mo_ket_noi += 1
        if not KetNoiCsdl.co_san:
            raise RuntimeError("không kết nối được CSDL")


class DichVuDatHangSai:
    def thuc_hien(self, ma_khach, tong_tien):
        KetNoiCsdl()                          # <- khởi tạo thẳng hạ tầng
        import time
        return DonHang("DH-X", ma_khach, tong_tien, int(time.time()))


# ---- Self-check ----
if __name__ == "__main__":
    # ---- 1. CON BUG: miền gọi thẳng hạ tầng -> KHÔNG TEST ĐƯỢC ----
    khong_test_duoc = False
    try:
        DichVuDatHangSai().thuc_hien("KH-01", 100_000)
    except RuntimeError:
        khong_test_duoc = True
    assert khong_test_duoc, "muốn test một luật nghiệp vụ, phải dựng cả CSDL trước"
    assert KetNoiCsdl.so_lan_mo_ket_noi == 1, "và mỗi lần chạy test là một lần mở kết nối"
    # Hệ quả dây chuyền: test chậm -> test giòn (hỏng vì CSDL, không phải vì bug) -> không
    # ai chạy test nữa -> không ai viết test nữa. Bắt đầu từ đúng một dòng khởi tạo hạ tầng
    # nằm sai chỗ. Và `time.time()` ở dòng dưới cũng là hạ tầng — nó làm kết quả không tất
    # định (bài 67).

    # ---- 2. LÕI ỨNG DỤNG CHẠY VỚI 0 HẠ TẦNG ----
    kho, bao = KhoTrongBoNho(), BaoGia()
    dv = DichVuDatHang(kho, bao, DongHoCoDinh(1_700_000_000))

    d = dv.thuc_hien("KH-01", 250_000)
    assert d.ma == "DH-1" and d.tao == 1_700_000_000, "kết quả TẤT ĐỊNH"
    assert len(kho.bang) == 1, "đơn đã được lưu"
    assert bao.da_bao == ["KH-01:đã tạo đơn DH-1"], "và khách đã được báo"
    assert KetNoiCsdl.so_lan_mo_ket_noi == 1, "0 kết nối CSDL nào được mở thêm"
    # Ba dòng dựng bối cảnh, không mock, không khung phần mềm, không tệp cấu hình.

    # ---- 3. ĐỔI BỘ NỐI: sửa ĐÚNG MỘT DÒNG, ở gốc lắp ráp ----
    kho_sql = KhoSql()
    d2 = DichVuDatHang(kho_sql, bao, DongHoCoDinh(1_700_000_000)).thuc_hien("KH-02", 300_000)
    assert d2.ma == "DH-1", "cùng logic nghiệp vụ, không sửa một chữ nào trong miền"
    assert len(kho_sql.cau_lenh) == 1 and kho_sql.cau_lenh[0].startswith("INSERT"), \
        "nhưng lần này nó sinh SQL"
    # "Đổi CSDL sau này" hiếm khi xảy ra thật, và đó KHÔNG phải lý do chính. Lý do chính
    # là: bộ nối thứ hai — bản trong bộ nhớ — cho phép TEST, và nó được dùng hằng ngày.

    # ---- 4. ĐIỀU CHỈ PYTHON LÀM ĐƯỢC: PHỤ THUỘC BẰNG 0 Ở CẢ HAI CHIỀU ----
    assert KhoDonHang not in KhoTrongBoNho.__mro__, "bộ nối KHÔNG kế thừa cổng"
    assert isinstance(kho, KhoDonHang), "nhưng nó VẪN thoả cổng — vì có đúng phương thức"
    assert isinstance(kho_sql, KhoDonHang), "cả hai bộ nối, không cái nào import cổng"
    # Ở Java/C++, bộ nối phải `implements KhoDonHang` — nghĩa là hạ tầng vẫn phải import
    # miền. Ở Python với `Protocol`, chiều phụ thuộc bằng 0 ở CẢ HAI phía.
    #
    # Hệ quả thực tế: một lớp có sẵn từ thư viện bên thứ ba — viết ra trước khi dự án của
    # bạn tồn tại — vẫn dùng làm bộ nối được, không cần lớp bọc:
    class LopThuVienBenThuBa:
        """Giả lập một lớp cache có sẵn. Nó chưa bao giờ nghe tên `KhoDonHang`."""
        def __init__(self):
            self._d = {}

        def tim_theo_ma(self, ma):
            return self._d.get(ma)

        def luu(self, d):
            self._d[d.ma] = d

    ngoai = LopThuVienBenThuBa()
    assert isinstance(ngoai, KhoDonHang), "lớp bên thứ ba thoả cổng mà không biết cổng tồn tại"
    d3 = DichVuDatHang(ngoai, bao, DongHoCoDinh(7)).thuc_hien("KH-03", 1_000)
    assert d3.tao == 7 and ngoai.tim_theo_ma("DH-1") is not None, "và nó chạy được ngay"
    # Cạm bẫy đi kèm, phải biết: `runtime_checkable` chỉ kiểm TÊN phương thức, KHÔNG kiểm
    # chữ ký. Một lớp có `luu(self)` không tham số vẫn qua được `isinstance`:
    class BoNoiSaiChuKy:
        def tim_theo_ma(self): ...            # thiếu tham số `ma`
        def luu(self): ...                    # thiếu tham số `d`

    assert isinstance(BoNoiSaiChuKy(), KhoDonHang), "isinstance vẫn nói ĐẠT — chỉ kiểm tên"
    no_luc_chay = False
    try:
        DichVuDatHang(BoNoiSaiChuKy(), bao, DongHoCoDinh(1)).thuc_hien("KH-X", 1)
    except TypeError:
        no_luc_chay = True
    assert no_luc_chay, "và nó chỉ nổ lúc CHẠY — nên `mypy` là bắt buộc, không phải tuỳ chọn"

    # ---- 5. CỔNG PHẢI NÓI TIẾNG NGHIỆP VỤ, KHÔNG NÓI TIẾNG BỘ NỐI ----
    # Cổng RÒ RỈ (rất hay gặp):
    #     class KhoDonHang(Protocol):
    #         def execute(self, sql: str) -> Cursor: ...
    # Nó "là Protocol" nên trông như đã đảo ngược phụ thuộc — nhưng chưa. Ba hậu quả:
    #   1. Không viết nổi bản trong bộ nhớ (lấy đâu ra `Cursor`?) -> mất cái lợi ở phần 3;
    #   2. Miền vẫn phải import driver CSDL -> chiều phụ thuộc vẫn ngược;
    #   3. Đổi sang kho khoá-giá trị là phải sửa cổng, tức là sửa miền.
    # Phép thử: đọc tên phương thức của cổng lên. Người làm nghiệp vụ hiểu được thì cổng
    # đúng; chỉ lập trình viên hiểu thì đó là bộ nối đội lốt cổng (bài 81).
    ten_pt = [t for t in vars(KhoDonHang) if not t.startswith("_")]
    assert sorted(ten_pt) == ["luu", "tim_theo_ma"], "cổng NHỎ: 2 phương thức, tiếng nghiệp vụ (bài 52)"

    # ---- 6. BÀI TEST KIẾN TRÚC: CHIỀU PHỤ THUỘC ----
    # Lục giác KHÔNG phải một cấu trúc thư mục. Đổi tên gói thành `domain/`,
    # `infrastructure/` mà `import` vẫn đi từ trong ra ngoài thì chẳng có gì thay đổi.
    # Luật thật chỉ có một: LÕI KHÔNG ĐƯỢC THAM CHIẾU HẠ TẦNG. Và nó kiểm được:
    HA_TANG = {KhoTrongBoNho, KhoSql, BaoGia, DongHoCoDinh, KetNoiCsdl}
    LOI = [DonHang, DichVuDatHang]
    vi_pham = [f"{lop.__name__}.{ten}"
               for lop in LOI for ten, gt in vars(lop).items()
               if gt in HA_TANG]
    assert vi_pham == [], f"0 tham chiếu từ lõi ra hạ tầng: {vi_pham}"
    # Trong dự án thật, bài kiểm tra mạnh hơn là quét `import` của gói `mien/`:
    #     grep -rE '^(from|import) (sqlalchemy|requests|redis|django)' mien/
    # Kết quả rỗng = đạt. Nó bắt đúng thời điểm ai đó "cho tiện" import một driver vào một
    # module miền — thời điểm kiến trúc bắt đầu tan rã, và mọi test nghiệp vụ vẫn xanh.

    # ---- 7. HAI LOẠI CỔNG, VÀ VÌ SAO PHẢI PHÂN BIỆT ----
    #
    #   Loại              | Ai gọi ai              | Ví dụ ở đây     | Bộ nối
    #   ------------------|------------------------|-----------------|------------------
    #   Cổng ĐIỀU KHIỂN   | thế giới -> miền       | `DatHang`       | REST, CLI, hàng đợi
    #   Cổng BỊ ĐIỀU KHIỂN| miền -> thế giới       | `KhoDonHang`    | CSDL, SMTP, đồng hồ
    #
    # Cả hai đều do MIỀN định nghĩa — điểm mấu chốt và cũng là chỗ hay sai. Với cổng bị
    # điều khiển thì ai cũng hiểu; với cổng điều khiển thì người ta hay để khung web định
    # nghĩa (view gọi thẳng vào lớp dịch vụ). Hậu quả: chữ ký use case bị định hình bởi
    # HTTP, và một job nền muốn dùng lại thì phải giả lập request.
    assert isinstance(dv, DatHang), "lõi thoả cổng điều khiển"
    assert len([t for t in vars(DatHang) if not t.startswith("_")]) == 1, \
        "cổng điều khiển = MỘT use case"

    # ---- 8. GỐC LẮP RÁP, VÀ KHI NÀO KHÔNG CẦN LỤC GIÁC ----
    # Có đúng một chỗ được phép khởi tạo cả miền lẫn hạ tầng — `main`, hoặc module cấu
    # hình. Mọi chỗ khác chỉ nhận phụ thuộc qua constructor (bài 51).
    #
    #   main() -> DichVuDatHang(KhoSql(), GuiEmailThat(), DongHoHeThong())
    #   test() -> DichVuDatHang(KhoTrongBoNho(), BaoGia(), DongHoCoDinh(42))
    #
    # Hai dòng đó là toàn bộ khác biệt giữa chạy thật và chạy test. Nếu để đổi sang test
    # bạn phải sửa file cấu hình hay đặt biến môi trường, thì gốc lắp ráp chưa tồn tại.
    #
    # Và lục giác có chi phí thật. Ba dấu hiệu ĐỦ để cần:
    #   - có luật nghiệp vụ đáng test riêng (không chỉ đọc/ghi bảng);
    #   - có nhiều hơn một đường vào (REST + hàng đợi + job nền);
    #   - có hệ ngoài mà bạn không kiểm soát (bài 94).
    # Thiếu cả ba thì một view gọi thẳng ORM là thiết kế đúng — và biết lúc nào KHÔNG áp
    # dụng một mẫu cũng là một phần của việc hiểu nó.
    assert DichVuDatHang(KhoTrongBoNho(), BaoGia(), DongHoCoDinh(42)).thuc_hien("KH-9", 1).tao == 42, \
        "lắp ráp cho test: một dòng"

    print("OK")

File: 98-hexagonal/HexagonalDemo.cpp
/*
 * Ngôn ngữ: C++
 * Công dụng: Bản C++ của cùng bài học — miền định nghĩa CỔNG, hạ tầng cung cấp BỘ NỐI,
 * mọi phụ thuộc chỉ đi vào TRONG. Ba con bug: miền gọi thẳng hạ tầng nên không test được;
 * cổng nói tiếng bộ nối; và "cấu trúc thư mục lục giác" mà chiều phụ thuộc vẫn ngược.
 * Tại sao cần học: ở Java và Python, "chiều phụ thuộc" là một luật về `import` mà ta phải
 * tự kiểm bằng test. Ở C++ nó có một phiên bản CỨNG HƠN và thấy được ngay: chiều của
 * `#include`. Header của miền không được `#include` bất cứ thứ gì thuộc hạ tầng — và nếu
 * ai đó vi phạm, cái giá không chỉ là kiến trúc mà là THỜI GIAN BUILD (bài 93). C++ còn
 * cho một lựa chọn thứ hai mà hai ngôn ngữ kia không có: cổng dưới dạng THAM SỐ TEMPLATE,
 * đảo ngược phụ thuộc mà không tốn một lời gọi ảo nào.
 */
#include <iostream>
#include <map>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>
#include <cstdlib>

// =====================================================================
// MIEN — khong biet CSDL, mang, dong ho hay khung phan mem nao ton tai
// (Trong du an that, toan bo phan nay nam trong `mien/` va header cua no
//  khong `#include` bat cu thu gi ngoai thu vien chuan.)
// =====================================================================
struct DonHang {
    std::string ma, maKhach;
    long tongTien, tao;
    DonHang(std::string m, std::string mk, long tt, long t)
        : ma(std::move(m)), maKhach(std::move(mk)), tongTien(tt), tao(t) {
        if (tongTien <= 0) throw std::invalid_argument("tong tien phai duong");
    }
};

// ---- CONG BI DIEU KHIEN (mien GOI RA ngoai) ----
// Chu y tu ngu: "tim theo ma", "luu", "bao cho khach". Khong co `sqlite3_stmt*`, khong
// co `CURL*`. Cong noi tieng NGHIEP VU (bai 81).
class KhoDonHang {
public:
    virtual ~KhoDonHang() = default;
    virtual const DonHang* timTheoMa(const std::string& ma) const = 0;
    virtual void luu(const DonHang& d) = 0;
};
class BaoChoKhach {
public:
    virtual ~BaoChoKhach() = default;
    virtual void bao(const std::string& maKhach, const std::string& noiDung) = 0;
};
class DongHo {
public:
    virtual ~DongHo() = default;
    virtual long bayGio() const = 0;      // bai 67
};

// ---- CONG DIEU KHIEN (the gioi GOI VAO mien) ----
class DatHang {
public:
    virtual ~DatHang() = default;
    virtual DonHang thucHien(const std::string& maKhach, long tongTien) = 0;
};

// Loi ung dung: cai cong dieu khien, dung cong bi dieu khien. 0 phu thuoc ha tang.
class DichVuDatHang : public DatHang {
public:
    DichVuDatHang(KhoDonHang& kho, BaoChoKhach& bao, const DongHo& dongHo)
        : kho_(kho), bao_(bao), dongHo_(dongHo) {}

    DonHang thucHien(const std::string& maKhach, long tongTien) override {
        DonHang d("DH-" + std::to_string(++dem_), maKhach, tongTien, dongHo_.bayGio());
        kho_.luu(d);
        bao_.bao(maKhach, "da tao don " + d.ma);
        return d;
    }
private:
    KhoDonHang& kho_;
    BaoChoKhach& bao_;
    const DongHo& dongHo_;
    int dem_ = 0;
};

// =====================================================================
// BO NOI — ha tang. Chung biet mien; mien KHONG biet chung.
// =====================================================================
class KhoTrongBoNho : public KhoDonHang {
public:
    const DonHang* timTheoMa(const std::string& ma) const override {
        auto it = bang.find(ma);
        return it == bang.end() ? nullptr : &it->second;
    }
    void luu(const DonHang& d) override { bang.insert({d.ma, d}); }
    std::map<std::string, DonHang> bang;
};

// Bo noi thu hai: gia lap SQL. Cung cong, cai dat hoan toan khac.
class KhoSql : public KhoDonHang {
public:
    const DonHang* timTheoMa(const std::string& ma) const override {
        cauLenh.push_back("SELECT * FROM don_hang WHERE ma = '" + ma + "'");
        auto it = bang.find(ma);
        return it == bang.end() ? nullptr : &it->second;
    }
    void luu(const DonHang& d) override {
        cauLenh.push_back("INSERT INTO don_hang VALUES ('" + d.ma + "', ...)");
        bang.insert({d.ma, d});
    }
    mutable std::vector<std::string> cauLenh;
    std::map<std::string, DonHang> bang;
};

class BaoGia : public BaoChoKhach {
public:
    void bao(const std::string& maKhach, const std::string& noiDung) override {
        daBao.push_back(maKhach + ":" + noiDung);
    }
    std::vector<std::string> daBao;
};
class DongHoCoDinh : public DongHo {
public:
    explicit DongHoCoDinh(long l) : luc_(l) {}
    long bayGio() const override { return luc_; }
private:
    long luc_;
};

// =====================================================================
// BAN SAI — mien goi thang ha tang
// =====================================================================
struct KetNoiCsdl {
    static int soLanMoKetNoi;
    static bool coSan;
    KetNoiCsdl() {
        ++soLanMoKetNoi;
        if (!coSan) throw std::logic_error("khong ket noi duoc CSDL");
    }
};
int KetNoiCsdl::soLanMoKetNoi = 0;
bool KetNoiCsdl::coSan = false;

struct DichVuDatHangSai {
    DonHang thucHien(const std::string& maKhach, long tongTien) {
        KetNoiCsdl kn;                                  // <- dung `new`/khoi tao thang ha tang
        return DonHang("DH-X", maKhach, tongTien, 0);
    }
};

// =====================================================================
// CONG DUOI DANG THAM SO TEMPLATE — dieu chi C++ co (phan 7)
// =====================================================================
template <class Kho, class Bao, class Ho>
class DichVuDatHangTinh {
public:
    DichVuDatHangTinh(Kho& k, Bao& b, const Ho& h) : kho_(k), bao_(b), ho_(h) {}
    DonHang thucHien(const std::string& maKhach, long tongTien) {
        DonHang d("DH-" + std::to_string(++dem_), maKhach, tongTien, ho_.bayGio());
        kho_.luu(d);
        bao_.bao(maKhach, "da tao don " + d.ma);
        return d;
    }
private:
    Kho& kho_;
    Bao& bao_;
    const Ho& ho_;
    int dem_ = 0;
};

// ---- Self-check ----
static void check(bool ok, const std::string& msg) {
    if (!ok) {
        std::cerr << "FAIL: " << msg << "\n";
        std::exit(1);
    }
}

int main() {
    // ---- 1. CON BUG: mien goi thang ha tang -> KHONG TEST DUOC ----
    bool khongTestDuoc = false;
    try { DichVuDatHangSai().thucHien("KH-01", 100000L); }
    catch (const std::logic_error&) { khongTestDuoc = true; }
    check(khongTestDuoc, "muon test mot luat nghiep vu, phai dung ca CSDL truoc");
    check(KetNoiCsdl::soLanMoKetNoi == 1, "va moi lan chay test la mot lan mo ket noi");
    // He qua day chuyen: test cham -> test gion (hong vi CSDL, khong phai vi bug) ->
    // khong ai chay test nua -> khong ai viet test nua. Bat dau tu dung mot dong khoi tao
    // ha tang nam sai cho.

    // ---- 2. LOI UNG DUNG CHAY VOI 0 HA TANG ----
    KhoTrongBoNho kho;
    BaoGia bao;
    DongHoCoDinh ho(1700000000L);
    DichVuDatHang dv(kho, bao, ho);

    DonHang d = dv.thucHien("KH-01", 250000L);
    check(d.ma == "DH-1" && d.tao == 1700000000L, "ket qua TAT DINH");
    check(kho.bang.size() == 1, "don da duoc luu");
    check(bao.daBao.size() == 1 && bao.daBao[0] == "KH-01:da tao don DH-1", "khach da duoc bao");
    check(KetNoiCsdl::soLanMoKetNoi == 1, "0 ket noi CSDL nao duoc mo them");
    // Bon dong dung boi canh, khong mock, khong khung phan mem, khong tep cau hinh.

    // ---- 3. DOI BO NOI: sua DUNG MOT DONG, o goc lap rap ----
    KhoSql khoSql;
    DichVuDatHang dvSql(khoSql, bao, ho);
    DonHang d2 = dvSql.thucHien("KH-02", 300000L);
    check(d2.ma == "DH-1", "cung logic nghiep vu, khong sua mot chu nao trong mien");
    check(khoSql.cauLenh.size() == 1 && khoSql.cauLenh[0].rfind("INSERT", 0) == 0,
          "nhung lan nay no sinh SQL");
    // "Doi CSDL sau nay" hiem khi xay ra that, va do KHONG phai ly do chinh. Ly do chinh
    // la: bo noi thu hai — ban trong bo nho — cho phep TEST, va no duoc dung hang ngay.

    // ---- 4. CONG PHAI NOI TIENG NGHIEP VU, KHONG NOI TIENG BO NOI ----
    // Cong RO RI (rat hay gap):
    //     class KhoDonHang { virtual sqlite3_stmt* query(const char* sql) = 0; };
    // No "la interface" nen trong nhu da dao nguoc phu thuoc — nhung chua. Ba hau qua:
    //   1. Khong viet noi ban trong bo nho (lay dau ra `sqlite3_stmt*`?) -> mat cai loi o
    //      phan 3;
    //   2. Header cua mien phai `#include <sqlite3.h>` -> chieu phu thuoc van nguoc, VA
    //      moi file cua mien phai dich lai khi thu vien do doi (bai 93);
    //   3. Doi sang kho khoa-gia tri la phai sua cong, tuc la sua mien.
    // Phep thu: doc ten phuong thuc cua cong len. Neu nguoi lam nghiep vu hieu duoc thi
    // cong dung; neu chi lap trinh vien hieu thi do la bo noi doi lot cong (bai 81).

    // ---- 5. CHIEU PHU THUOC — o C++ no la chieu cua `#include` ----
    // Luc giac KHONG phai mot cau truc thu muc. Doi ten thu muc thanh `domain/`,
    // `infrastructure/` ma `#include` van di tu trong ra ngoai thi chang co gi thay doi.
    // Luat that chi co mot: LOI KHONG DUOC THAM CHIEU HA TANG.
    //
    // O C++, luat do kiem duoc bang mot lenh, khong can thu vien nao:
    //     grep -rE '#include\s*[<"](sqlite3|curl|pqxx|boost/asio)' mien/
    // Ket qua rong = dat. Va no manh hon phien ban Java/Python o mot diem: `#include` la
    // phu thuoc VAT LY, nen vi pham khong chi lam ban kien truc — no lam moi file cua mien
    // phai dich lai khi thu vien ha tang doi mot dong.
    KhoDonHang* conTro = &khoSql;
    check(conTro != nullptr, "ha tang CAI cong cua mien — mui ten chi vao trong");
    check(dynamic_cast<KhoDonHang*>(&kho) != nullptr, "ca hai bo noi deu la KhoDonHang");

    // ---- 6. HAI LOAI CONG, VA VI SAO PHAI PHAN BIET ----
    //
    //   Loai              | Ai goi ai              | Vi du o day     | Bo noi
    //   ------------------|------------------------|-----------------|------------------
    //   Cong DIEU KHIEN   | the gioi -> mien       | `DatHang`       | REST, CLI, hang doi
    //   Cong BI DIEU KHIEN| mien -> the gioi       | `KhoDonHang`    | CSDL, SMTP, dong ho
    //
    // Ca hai deu do MIEN dinh nghia — do la diem mau chot va cung la cho hay sai. Voi cong
    // bi dieu khien thi ai cung hieu; voi cong dieu khien thi nguoi ta hay de khung web
    // dinh nghia. Hau qua: chu ky use case bi dinh hinh boi HTTP, va mot job nen muon dung
    // lai thi phai gia lap request.
    DatHang* congDieuKhien = &dv;
    check(congDieuKhien->thucHien("KH-03", 1L).maKhach == "KH-03", "goi qua cong, khong goi thang lop");

    // ---- 7. DIEU CHI C++ CO: CONG DUOI DANG THAM SO TEMPLATE ----
    // Cung nguyen tac dao nguoc phu thuoc, nhung quyet dinh luc BIEN DICH: khong bang ao,
    // khong con tro, noi tuyen duoc — dung nhu bai 88 phan 5.
    KhoTrongBoNho kho2;
    BaoGia bao2;
    DongHoCoDinh ho2(99L);
    DichVuDatHangTinh<KhoTrongBoNho, BaoGia, DongHoCoDinh> dvTinh(kho2, bao2, ho2);
    DonHang d3 = dvTinh.thucHien("KH-04", 500L);
    check(d3.tao == 99L && kho2.bang.size() == 1, "cung ket qua, khong mot loi goi ao nao");
    //
    //   Cong ao (`virtual`)        | Cong template
    //   ---------------------------|---------------------------------
    //   chon bo noi LUC CHAY       | chon LUC BIEN DICH
    //   mot ban ma duy nhat        | moi to hop sinh mot ban ma rieng
    //   doc duoc tu cau hinh       | KHONG
    //   co chi phi goi ao          | 0
    //
    // Quy tac chon giong bai 88: bo noi den TU CAU HINH (doi CSDL theo moi truong) thi
    // dung cong ao; bo noi co dinh va nam trong vong lap nong thi dung template. Voi phan
    // lon ung dung nghiep vu, cong ao la lua chon dung — chi phi goi ao khong dang ke so
    // voi mot luot truy van CSDL.

    // ---- 8. GOC LAP RAP, VA KHI NAO KHONG CAN LUC GIAC ----
    // Co dung mot cho trong chuong trinh duoc phep khoi tao ca mien lan ha tang — `main`.
    // Moi cho khac chi nhan phu thuoc qua constructor (bai 51).
    //
    // Va luc giac co chi phi that: moi cong la mot lop truu tuong, moi bo noi mot lop. Voi
    // mot ung dung CRUD thuan thi do la ba lop cho mot viec ma `save()` lam xong. Ba dau
    // hieu DU de can:
    //   - co luat nghiep vu dang test rieng (khong chi doc/ghi bang);
    //   - co nhieu hon mot duong vao (REST + hang doi + job nen);
    //   - co he ngoai ma ban khong kiem soat (bai 94).
    // Thieu ca ba thi mot controller goi thang repository la thiet ke dung.
    check(kho.bang.size() == 2 && khoSql.bang.size() == 1 && kho2.bang.size() == 1,
          "ba bo noi da chay, mot mien duy nhat — va mien khong biet cai nao la cai nao");

    std::cout << "OK\n";
    return 0;
}
